dm-source-code-reading-4.md

```diff
@@ -0,0 +1,156 @@
+
+需要注意目前 DM 在任务配置中指定的库表黑白名单功能只应用于 load 和 binlog replication 处理单元。如果在 dump 处理单元内使用库表黑白名单功能，需要在同步任务配置文件的 dump 处理单元配置提供 extra-args 参数，并指定 mydumper 相关参数，包括 --database, --tables-list 和 --regex。mydumper 使用 regex 过滤库表的实现参考 [`check_regex`](https://github.com/pingcap/mydumper/blob/9493dd752b9ea8804458e56a955e7f74960fa969/mydumper.c#L314-L338) 函数。
+
+## load 处理单元
+
+load 处理单元的代码位于 [github.com/pingcap/dm/loader](https://github.com/pingcap/dm/tree/master/loader) 包内，该处理单元在 dump 处理单元运行结束后运行，读取 dump 处理单元导出的 SQL 文件解析并在下游数据库执行逻辑 SQL。我们重点分析 `Init` 和 `Process` 两个 interface 的实现。
+
+### Init 实现细节
+
+该阶段进行一些初始化和清理操作，并不会开始同步任务，如果在该阶段运行中出现错误，会通过 [rollback 机制](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L356-L361) 清理资源，不需要调用 Close 函数。该阶段包含的初始化操作包括以下几点：
+
+* [创建 `checkpoint`](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L363)，`checkpoint` 用于记录全量数据的导入进度和 load 处理单元暂停或异常终止后，恢复或重新开始任务时可以从断点处继续导入数据。
+
+* 应用任务配置的数据同步规则，包括以下规则：
+    * [初始化黑白名单](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L370)
+    * [初始化表路有规则](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L380)
+    * [初始化列值转换规则](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L385-L390)
+
+### Process 实现细节
+
+该阶段的工作流程也很直观，通过 [一个收发数据类型为 `*pb.ProcessError` 的 `channel` 接收运行过程中出现的错误，出错后通过 context 的 `CancelFunc` 强制结束处理单元运行](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L408-L422)。在核心的 [数据导入函数](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L485) 中，工作模型与 mydumper 类似，即在 [主线程中分发任务](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L507)，[有多个工作线程执行具体的数据导入任务](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L500-L503)。具体的工作细节如下：
+
++ 主线程会按照库，表的顺序读取创建库语句文件 `<db-name>-schema-create.sql` 和建表语句文件 `<db-name>.<table-name>-schema-create.sql`，并在下游执行 SQL 创建相对应的库和表。
+
++ [主线程读取 `checkpoint` 信息，结合数据文件信息创建 fileJob 随机分发任务给一个工作子线程](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L944-L1015)，fileJob 任务的结构如下所示	：
+	
+	```
+	type fileJob struct {
+	   schema    string
+	   table         string
+	   dataFile   string
+	   offset       int64            // 表示读取文件的起始 offset，如果没有 checkpoint 断点信息该值为 0
+	   info           *tableInfo  // 保存原库表，目标库表，列名，insert 语句 column 名字列表等信息
+	}
+	```
+
++ 在每个工作线程内部，有一个循环不断从自己 `fileJobQueue` 获取任务，每次获取任务后会对文件进行解析，并将解析后的结果分批次打包为 SQL 语句分发给线程内部的另外一个工作协程，该工作协程负责处理 SQL 语句的执行。工作流程的伪代码如下所示，完整的代码参考 [`func (w *Worker) run()`](https://github.com/pingcap/dm/blob/25f95ee08d008fb6469f0b172e432270aaa6be52/loader/loader.go#L114-L173)：
+
+	```
+	// worker 工作线程内分发给内部工作协程的任务结构
+	type dataJob struct {
```

Comment:
It seems the variable fields are not alignment, I'm not sure whether we use space or tab here, it is better to make them align.